### ISSUE TITLE
Prevent wrapping/shrinking of specific columns when using `#pack`

### ISSUE THREAD

**Issue by jaynetics** [matt-harvey/tabulo#18]
hi,

nice gem! thanks for making it!

i have a use case where i'd like to use `#pack`, but prevent wrapping for one column.

this is for the gem [invalid_record_finder]( https://github.com/betterplace/invalid_record_finder ), which finds broken rails records and reports them. one output format uses `tabulo`. sometimes the records have UUIDs, and if these UUIDs wrapped, its a pain to copy-paste them from the table.

example:

```ruby
entry = {
  model:       'BlogPost',
  field:       'body_locale_en',
  field_value: 'hello! wassup?',
  message:     'Your input is too short (less than 20 chr)',
  id:          '36e8f2361c704977b06ab37a96901215',
}
table = Tabulo::Table.new([OpenStruct.new(entry)], *entry.keys)
puts table.pack.to_s
```

the output i get:

```
+----------+----------------+----------------+----------------+----------------+
|   model  |      field     |   field_value  |     message    |       id       |
+----------+----------------+----------------+----------------+----------------+
| BlogPost | body_locale_en | hello! wassup? | Your input is  | 36e8f2361c7049 |
|          |                |                | too short, les | 77b06ab37a9690 |
|          |                |                | s than 20 chr) | 1215           |
+----------+----------------+----------------+----------------+----------------+
```

the output i would like to achieve (id column doesn't wrap so it can be easily copy-pasted):

```
+----------+----------+----------+----------+----------------------------------+
|   model  |   field  | field_va |  message |                id                |
|          |          |    lue   |          |                                  |
+----------+----------+----------+----------+----------------------------------+
| BlogPost | body_loc | hello! w | Your inp | 36e8f2361c704977b06ab37a96901215 |
|          | ale_en   | assup?   | ut is to |                                  |
|          |          |          | o short  |                                  |
|          |          |          | (less th |                                  |
|          |          |          | an 20 ch |                                  |
|          |          |          | r)       |                                  |
+----------+----------+----------+----------+----------------------------------+
```

some ideas on how to achieve this:

- add `Column#min_width` (optional Integer, default = 1)
  - i have implemented this in PR #17
  - good:
    - simple implementation
    - could be re-used for other use cases
  - bad:
    - not obvious that it only affects `#pack` output
    - cuts across concerns a bit (the column doesn't care about its own `min_width` internally)
    - leaves calculation of desired `min_width` to the user (`entries.map { |e| e.id.length }.max`)
- add `Column#wrap_on_pack` (optional Boolean, default = true)
  - good:
    - makes it more explicit that it only affects `#pack` output
  - bad:
    - seems like more code needs to be touched to implement this
    - cuts across concerns a bit
    - hard to find a clear, non-obtrusive name?
- add a keyword to `#pack`, something like `no_shrink_columns:` (optional Array, default = [])
  - good:
    - makes it most explicit that it only affects `#pack` output
  - bad:
    - seems to me to require the biggest amount of changes
    - hard to find a clear, non-obtrusive name?

*edit:* ... or maybe i missed a way that already allows doing this?!

**Comment by matt-harvey** [comment]
Hi @jaynetics , glad you like Tabulo! Many thanks for the PR(s), and for this Issue report with its thoughtful analysis of alternative solutions.

OK, so, firstly, it is actually possible to achieve your desired outcome right now, without any changes to Tabulo. The key is to construct the table first _sans_ the columns you want not to shrink; then `.pack` the table to a width that allows enough wriggle room for adding the columns later; then add the non-shrinking columns. If you want the end result to be "packed" according to your screen width, then you can use the [unicode-display_width](https://rubygems.org/gems/unicode-display_width) gem to calculate your screen width before subtracting the "wriggle room". Here's how that would look:

```rb
require 'tabulo'
require 'ostruct'
require 'tty-screen'

entry = {
  model:       'BlogPost',
  field:       'body_locale_en',
  field_value: 'hello! wassup?',
  message:     'Your input is too short (less than 20 chr)',
  id:          '36e8f2361c704977b06ab37a96901215',
}
id_column_width = entry[:id].length
padding_and_borders_width_for_id_column = 3
wriggle_room = id_column_width + padding_and_borders_width_for_id_column

table = Tabulo::Table.new([OpenStruct.new(entry)], :model, :field, :field_value, :message)
final_table_width = TTY::Screen.width  # or 80, or whatever you want it to be
table.pack(max_table_width: final_table_width - wriggle_room)
table.add_column(:id, width: id_column_width)  # use before: option if you want to put it somewhere other than the end
puts table  # note '.to_s' is unnecessary
```

So this will give you want you want right now.

---

Having said that, this workaround is admittedly awkward. It would be worthwhile to have a feature where columns can be marked as "unshrinkable" on calling `.pack`. I think my preferred solution here would be to add an option to `.pack`. Perhaps the following:

```rb
table.pack(except: [:id, :foo])

We could also provide its dual:

```rb
table.pack(only: [:bar, :baz])
```

Importantly, the option should receive [column labels](https://github.com/matt-harvey/tabulo#column-labels-vs-headers-) and not numeric indices.

I think this could be implemented simply enough.

The naming pattern of "only/except" is also found in Rails, so there's a little bit of precedent there. (Rails also allows single symbols to be passed as well as arrays, e.g. `except: :id`.) I like that this option makes it really clear that this has everything to do with the `pack` operation, and nothing at all to do with the column _in itself_. Conversely, having a `min_width` option on the column itself could cause subtle confusions.

What do you think?

---

Just as an aside, in case you're not aware of it, there is an option to [preserve whole words](https://github.com/matt-harvey/tabulo#preserve-words) when wrapping, _if possible_. You might find this useful for your `message` column to aid in readability of the output. This particular option won't actually help you with the `id` column, because it's not taken into account by the `pack` algorithm. (To adjust the algorithm to take it into account would have been quite complicated, I think.) However, I just wanted to mention it tangentially, as you might find it useful.

**Comment by jaynetics** [comment]
thanks for the detailed feedback!

just doing some calculating myself is what i came up with initially. after all, it isn't super hard to do it just for a single case.

it just felt like re-implementing part of the functionality that is already in the gem.

`only:` and `except:` keywords sound nice to me. the usage in my case would be something like this:

```ruby
table = Tabulo::Table.new(data, :foo, :bar) do |t|
  t.add_column(:id, width: data.map { |el| el.id.length }.max)
end
table.pack(except: :id)
```

this would work and isn't too complicated, in terms of either implementation or usage.

i'm just wondering whether it might make sense to do the max-length calculation on the side of the library, as it already goes through all the data anyway.

a parameter to trigger that could be called something like `nowrap:` and be used like so:

```ruby
table = Tabulo::Table.new(data, :foo, :bar, :id)
table.pack(nowrap: :id)
```

the way i imagine it, this would have the same result as the example above. internally, it would use the non-wrapped length of the `header` and `formatted_content` for setting the width of the `id` column in `pack`, and it would then exclude that column from shrinking in `shrink_to`.

possible values for `nowrap` could be `nil` (default), `true` (prevent wrapping for all columns), or one or more column names.

for my case, this would be the most convenient interface, but you'll have to decide what is most likely to be useful for the target audience of this gem.

i'm happy to prepare a PR either way.

**Comment by matt-harvey** [comment]
> i'm just wondering whether it might make sense to do the max-length calculation on the side of the library, as it already goes through all the data anyway

Good point. Making the user do their own width calculation doesn't feel great. One of the core purposes of Tabulo is to do those calculations for you.

I think the key here is that `pack` is actually a convenience function around what are two (actually three if you count the title adjustment, but let's gloss over that for now) distinct operations: autosizing the columns based on their contents; and shrinking the table to a desired overall width.

So there are two approaches that now spring to mind:
1. Provide a method `.autosize_columns`, and method `.shrink_to`, as public, "plumbing" methods, while retaining `pack` as the "porcelain" method.  All three methods could accept an `except` parameter (and perhaps also an `only` parameter, although I'm not convinced of its value just yet). Thus, a user for whom `pack` does not quite give them what they want, could drop down to the plumbing methods and chain them together to achieve the desired result. I.e. `table.autosize_columns.shrink_to(:screen, except: [:id])`.
1. Embrace the fact that the `pack` is a convenience method that does conceptually diverse things for the sake of user convenience, to give them an end result that is most likely to look decent on the page, and provide a `nowrap` option also for the sake of user convenience for what is probably the most likely use case for wanting to modify the default behaviour.

The advantage of (1) is that it introduces a keyword that can have multiple "orthogonal" uses; applying to any method that takes action upon all the table's columns, and then clearly stating: "let's not apply _any_ of this method to _any_ of these columns". It also provides finer grained control over the various adjustments by exposing the "plumbing" behind `pack` to the user. The disadvantages of (1) are that it exposes two more methods that will need to be documented and maintained, i.e. it increases the surface area of the API, and that it will might be harder for users to discover exactly how they're supposed to chain these methods together to achieve their desired outcome.

The advantage of (2) is convenience. This is probably the most common use case for modulating `pack` that's not already provided.

One thing about (2) that I'm unsure about is, the name of the option. `nowrap` has precedent in CSS: AFAIA, it means _never wrap_. However, its proposed use in Tabulo wouldn't strictly mean that. A cell that was already wrapped because it contains `\n` or because the user already set its width to something manually, that already meant it was wrapped, would still be wrapped even after `pack`ing with `nowrap`. So maybe it should be called `noshrink`? But maybe that sounds kind of weird...?

Also just as an aside, I don't think the `nowrap: true` option (as opposed to `nowrap: [:foo, :bar]`) adds much value. If the user wants to avoid wrapping any columns at all, they can already achieve this by calling `.pack(max_table_width: nil)`. Unless I'm missing something?

Given all of the above, I'm honestly still undecided on the best way forward here.

In any case: thank you for your offer of a PR; I would really appreciate that! 😃 It's just that I would like to mull it over a bit more before expanding the API in a particular direction, is all. If you have further thoughts on any of these or other options, they are most welcome.

**Comment by matt-harvey** [comment]
Hey @jaynetics , the more I think about this, the more I think that the best course of action for now, is to provide an `except:` parameter on the `.pack` option, and to have this take an array of symbols, being column labels, and then make it so that the `.pack` method will not change (neither increase, nor decrease) the width of any of the columns thus identified. To me, this solution is:
* Straightforward to communicate clearly in the documentation what the behaviour of it is (`pack` _will not affect_ these columns, period).
* "Forwards compatible": The `except:` concept can be applied to other cases in future without adding further "concepts" to the public API. It's nicely orthogonal to the further customisation that may one day be exposed by the "porcelain" methods referred to in my previous comment.
* It addresses the original issue well _enough_, allowing users to achieve the "don't wrap this column" behaviour without _too_ much trouble.

In regards to the original issue, I suspect that for many of the use cases, when a user definitely wants not to wrap, they will know the width of the column in advance anyway. For example, if they have a UUID column, they will know the width of a UUID without having to traverse the collection. So to me, the case for `nowrap` feels a little bit too specific to the exact use issue that has arisen here. In contrast, the case for `except` is more generally applicable as a concept&dash;it feels more intuitively right to me as something I would be comfortable committing to as the API evolves.

Would you be happy to create a PR to add the `except:` functionality to `.pack` on this basis?

If you have reservations about this proposal, I'm still happy to consider alternatives; but I just wanted to let you know where my thinking is at at the moment.

Thanks again.

**Comment by jaynetics** [comment]
> I suspect that for many of the use cases, when a user definitely wants not to wrap, they will know the width of the column in advance anyway. For example, if they have a UUID column, they will know the width of a UUID without having to traverse the collection.

which use cases do you have in mind?

i tried to come up with some cases where people would want to avoid wrapping, and i think they might all require traversing the collection to find the maximum length:

- other data where copy & paste is useful, such as email addresses or phone numbers
- languages where extra line breaks could be confusing or affect meaning (e.g. Korean)
- visually structured data within columns - maybe even through recursive usage of `tabulo` (😍 )
- other visual structures such as ascii art

in my case, i actually need to traverse the collection as well. this is because the data can contain records from multiple DB tables, and i don't know whether any of these will use UUIDs. they might all just have Integer ids of arbitrary length.

i strongly agree with your analysis of the downsides of both options, though, and the `except:` solution is fine for me, so i prepared #21. i still have two small ideas, though:

option (1b): like option (1) but without detailed documentation. i agree the documentation is already quite expansive for this type of project, but just mentioning these methods might suffice for the few interested people to go look at the code and figure it out. just like option (1), this might have the added benefit of cutting `pack` into smaller parts which are easier to maintain.

option (3): a new value for `wrap_preserve` such as `:orginal`? not perfectly clean, because it would be "stronger" than the other two values in so far as it would be the only value to affect `shrink_to`, but arguably the most lightweight when it comes to introducing new concepts or enlarging the API.

(don't worry about the PR if these sound convincing to you, it was surprisingly little work).

**Comment by matt-harvey** [comment]
> which use cases do you have in mind?

UUIDs and nanoids mostly. (I said "many", not "most" 😜 .)

Remember also, though, that in order for this issue to arise at all, the table has to be wider than the screen width (or other configured total width) in the first place. The default behaviour is for the table to size itself so that _nothing_ is wrapped (except for cells with line breaks in the content). Then, if it does need to wrap to fit the screen, it first targets the widest column, then the next widest etc.. So a column has to be quite "unlucky" to be wrapped at all. And of these, there's then a still further subset where the user can't reasonably either know or estimate a workable fixed width up front.

> in my case, i actually need to traverse the collection as well.

I suspected as much, since your library can't make too many assumptions about the data it's dealing with.

I'm pretty sure that I want to expose those "plumbing" methods at some point&mdash;either in the next minor release, or the one after&mdash;so I think that will address that use case pretty well.

> just mentioning these methods might suffice for the few interested people to go look at the code and figure it out

I agree with just mentioning, rather than detailing, such methods in the README.

My preferred approach for documenting low-level knobs and levers work, is to very briefly mention their existence in the README, together with a link to the appropriate anchor tag(s) within the detailed, "formal" API documentation that lives [here](https://www.rubydoc.info/gems/tabulo/Tabulo).

Re. option 3: That's an interesting idea. I need to give it some further though. But as a mechanism for modulating the action of `.pack`, I still prefer `except:`.

**Comment by matt-harvey** [comment]
This is now addressed on `v2-8` branch, thanks to https://github.com/matt-harvey/tabulo/pull/21. I will close this issue after I've released `v2.8.0`.

**Comment by matt-harvey** [comment]
This has now been shipped in `v2.8.0`.
